compiler/infra/ForwardList.hpp

```diff
@@ -0,0 +1,684 @@
+#ifndef FORWARD_LIST_HPP
```

Comment:
TR::reference_wrapper is a re-implementation of std::reference_wrapper [I.E. it should be drop-in replaceable with std::reference_wrapper] that is sufficient for our current requirements.